File: l2/bios/bios.go
package bios
import (
	"luna_l2/shared"
	"luna_l2/proxy"
	"time"	
	"os"
	"fmt"
)

const (
	MEMSIZE uint32 = 0x70000000
	MEMCAP uint32 = 0x6FFFFFFF
)

// Interrupt modes:
	// 0: BIOS handle
	// 1: Software handle
	// 2: Dual handle (BIOS first)
	// 3: Dual handle (Software first)

func WriteChar(char string, fg uint8, bg uint8) {
	proxy.VideoPrintChar(rune(char[0]), byte(fg), byte(bg))
}

func WriteString(str string, fg uint8, bg uint8) {
	for _, r := range str {
		WriteChar(string(r), fg, bg)
	}
}

func WriteLine(str string, fg uint8, bg uint8) {
	WriteString(str + "\n", fg, bg)
}

func LoadSector(drive int, sector int, dest_sector int) bool {	
	var file string
	switch drive {
	case 0:
		file = shared.Filename
		time.Sleep(time.Duration(12) * time.Millisecond)
	case 1:
		file = shared.SDFilename
		time.Sleep(time.Duration(2) * time.Millisecond)
	case 2:
		file = shared.OpticalFilename
		time.Sleep(time.Duration(110) * time.Millisecond)
	}	
	f, err := os.OpenFile(file, os.O_RDONLY, 0)
	if err != nil {
		shared.SetRegister(0x0000, 1)
		fmt.Println("luna-l2: could not load/reload block device: ", err)
		return false
	}
	defer f.Close()
	start := sector * 512
	rstart := dest_sector * 512
	_, err = f.ReadAt((*shared.Memory)[start:start + 512], int64(rstart))
	if err != nil && err.Error() != "EOF" {
		shared.SetRegister(0x0000, 1)
		fmt.Println("luna-l2: could not read from disk: ", err)
	} else {
		shared.SetRegister(0x0000, 0)
	}
	return true
}

func WriteSector(drive int, sector int, dsector int) {
	var file string
	switch drive {
	case 0:
		file = shared.Filename
		time.Sleep(time.Duration(15) * time.Millisecond)
	case 1:
		file = shared.SDFilename
		time.Sleep(time.Duration(8) * time.Millisecond)
	case 2:
		file = shared.OpticalFilename
		time.Sleep(time.Duration(200) * time.Millisecond)
	}	
	
	f, err := os.OpenFile(file, os.O_RDWR | os.O_SYNC, 0)
	if err != nil {
		shared.SetRegister(0x0000, 1)
		fmt.Println("luna-l2: could not load/reload block device: ", err)
		return
	}
	defer f.Close()

	start := sector * 512
	_content_start := dsector * 512
	_, err = f.WriteAt((*shared.Memory)[start:start + 512], int64(_content_start))
	if err != nil {
		shared.SetRegister(0x0000, 1)
		fmt.Println("luna-l2: could not write to block device: ", err)
	} else {
		shared.SetRegister(0x0000, 0)
	}
}

func IntHandler(code uint32) {
	switch code {
	case 0x01:
		char := shared.GetRegister(0x0001)
		WriteChar(string(rune(char)), uint8(shared.GetRegister(0x0002)), uint8(shared.GetRegister(0x0003)))
	case 0x02:
		// Programmable interval timer reserved
	case 0x03:	
		// Disk inserted query
		var _file string
		switch shared.GetRegister(0x0001) {
		case 0:
			_file = shared.Filename
		case 1:
			_file = shared.SDFilename
		case 2:
			_file = shared.OpticalFilename
		}

		if _file != "" {
			shared.SetRegister(0x0001, 1)
		} else {
			shared.SetRegister(0x0001, 0)
		}
	case 0x04:
		// Syscall interrupt
	case 0x05:
		// Keyboard reserved
	case 0x06:
		// Power interrupt
	case 0x07:
		WriteLine("Illegal instruction 0x" + fmt.Sprintf("%08x", shared.GetRegister(0x0001)) + " at location 0x" + fmt.Sprintf("%08x", shared.GetRegister(0x001a)), 255, 0)
	case 0x08:
		// Unmapped
	case 0x09:
		// Unmapped
	case 0x0A:
		if shared.Bits32 == false {
			shared.SetRegister(0x0001, 0xffff)
		} else {
			shared.SetRegister(0x0001, MEMSIZE)
		}	
	case 0x0B:
		sector := shared.GetRegister(0x0001)
		drive := shared.GetRegister(0x0002)
		rsector := shared.GetRegister(0x0003)
		LoadSector(int(drive), int(sector), int(rsector))
	case 0x0C:
		proxy.VideoSetCursor(int(shared.GetRegister(0x0001)), int(shared.GetRegister(0x0002)))
	case 0x0D:
		sector := shared.GetRegister(0x0001)
		drive := shared.GetRegister(0x0002)
		rsector := shared.GetRegister(0x0003)
		WriteSector(int(drive), int(sector), int(rsector))
	case 0x0E:
		x, y := proxy.VideoGetCursor()
		shared.SetRegister(0x0001, uint32(x))
		shared.SetRegister(0x0002, uint32(y))
	case 0x0F:
		shared.BootDrive = int(shared.GetRegister(0x0001))
		for i, _ := range (*shared.Registers) {
			(*shared.Registers)[i].Value = uint32(0)
		}
		shared.Bits32 = false
		for i := 0; i < len((*shared.Memory)); i++ {
			(*shared.Memory)[i] = 0x00
		}
		proxy.VideoClearVideoMemory()
		proxy.VideoSetCursor(0, 0)
	case 0x10:
		shared.SetRegister(0x0001, uint32(shared.DriveNumber))
	case 0x11:
		// Shut down machine
	}	
}

func Splash() {
	WriteLine("Luna L2", 255, 0)
	WriteLine("BIOS: Integrated BIOS", 255, 0)	
	WriteLine("Copyright (c) 2025 zoeyfdev\n", 255, 0)
}

func CheckArgs() bool {
	if len(os.Args) < 2 {
		Splash()
		WriteLine("No bootable device", 255, 0)
		return false
	}
	return true
}

func IntWrapper(code uint32, next uint32) {
	// shared.SetRegister(0x001e, code)
	var mem_location uint32 = 0x6FFF0000 + uint32(((code - 1) * 6))	
	loc := uint32((*shared.Memory)[mem_location + 2]) << 24 | uint32((*shared.Memory)[mem_location + 3]) << 16 | uint32((*shared.Memory)[mem_location + 4]) << 8 | uint32((*shared.Memory)[mem_location + 5])
	
	switch (*shared.Memory)[mem_location + 1] {
	case 0x00:
		IntHandler(code)
	case 0x01:
		shared.SetRegister(0x001e, next)
		shared.SetRegister(0x001d, loc)
	case 0x02:
		IntHandler(code)
		shared.SetRegister(0x001e, next)
		shared.SetRegister(0x001d, loc)
	case 0x03:
		shared.SetRegister(0x001e, next)
		shared.SetRegister(0x001d, loc)
		IntHandler(code)
	}
}

File: l2/bios/bios.c
#include <stdint.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "../hardware/video_common.h"
#include "../cpu/registers.h"
#include "../cpu/memory.h"
#include "../cpu/memdefs.h"
#include "disk.h"

char* HDD_FILE;
char* SD_FILE;
char* DVD_FILE;
int bios_boot_drive;
unsigned char* BIOS_RAM = NULL;

#define BIOS_RAM_SIZE 192

void bios_write_string(char* s) {
    while (*s) {
        v_print_char(*s, 255, 0);
        s++;
    }
}

void bios_write_line(char* s) {
    bios_write_string(s);
    v_print_char(0x0a, 255, 0);
}

void bios_splash() {
    bios_write_line("Luna L2");
    bios_write_line("BIOS: Integrated BIOS");
    bios_write_line("Copyright (c) 2026 zoeyfdev\n");
}

void bios_init() {
    BIOS_RAM = malloc(BIOS_RAM_SIZE); // 6 bytes per interrupt * 32 interrupts
}

unsigned char bios_read_memory(uint32_t address) {
    if (address < BIOS_RAM_SIZE)
        return BIOS_RAM[address];
    return (unsigned char) rand() & 0xFF;
}

void bios_write_memory(uint32_t address, unsigned char value) {
    if (address < BIOS_RAM_SIZE)
        BIOS_RAM[address] = value;
}

void bios_handle_interrupt(uint32_t code) {
    uint32_t handler_addr = (code - 1) * 6;

    // 0: BIOS
    // 1: software
    // else: disabled
    if (BIOS_RAM[handler_addr] == 1) {
        uint32_t pc_addr = (BIOS_RAM[handler_addr + 1] << 24)
            | (BIOS_RAM[handler_addr + 2] << 16)
            | (BIOS_RAM[handler_addr + 3] << 8)
            | (BIOS_RAM[handler_addr + 4]);
        set_register(PC, pc_addr);
        return;
    }

    switch (code) {
    case 0x01:
        v_print_char((unsigned int) get_register(R1) & 0xFF, (unsigned int) get_register(R2) & 0xFF, (unsigned int) get_register(R3) & 0xFF);
        break;
    case 0x02:
        // PIT reserved
        break;
    case 0x03: {
            char* file;
            switch (get_register(R1)) {
            case 0:
                file = HDD_FILE;
                break;
            case 1:
                file = SD_FILE;
                break;
            case 2:
                file = DVD_FILE;
                break;
            }

            if (file != NULL)
                set_register(R1, 1);
            else
                set_register(R1, 0);
            break;
        }
    case 0x04:
        // Syscall reserved
        break;
    case 0x05:
        // Power reserved
        break;
    case 0x06:
        // Keyboard interrupt
        break;
    case 0x07:
        // Illegal instruction trap
        break;
    case 0x08:
        // Unmapped
        break;
    case 0x09:
        // Unmapped
        break;
    case 0x0A:
        // Memory query
        break;
    case 0x0B:
        // Load sector from disk
        load_sector(get_register(R2), get_register(R1), get_register(R3));
        break;
    case 0x0C:
        v_set_cursor(get_register(R1), get_register(R2));
        break;
    case 0x0D:
        write_sector(get_register(R2), get_register(R1), get_register(R3));
        break;
    case 0x0E: {
            int x;
            int y;
            v_get_cursor(&x, &y);

            set_register(R1, x);
            set_register(R2, y);
            break;
        }
    case 0x0F: {
            // Reboot
            memset(MEMORY, 0x00, MEMSIZE);
            memset(BIOS_RAM, 0x00, BIOS_RAM_SIZE);
            for (int i = 0; i < 35; i++)
                set_register(i, 0);
            v_gpu_reset();
            break;
        }
    case 0x10:
        set_register(R1, bios_boot_drive);
        break;
    case 0x11:
        exit(0);
        // Shut down machine
        break;
    }
}
